File: dataset/ZeroShot/GEMINI_gemini-2.5-pro/Task147.py
import heapq

def find_kth_largest(nums, k):
    """
    Finds the kth largest element in a list of integers using a min-heap.

    Args:
        nums: The input list of integers.
        k: The 'k' value, representing the desired largest element rank.

    Returns:
        The kth largest element.

    Raises:
        ValueError: If the input is invalid (e.g., empty list or k is out of bounds).
    """
    # 1. Input Validation
    if not nums or k < 1 or k > len(nums):
        raise ValueError("Invalid input: list is empty or k is out of bounds.")

    # 2. Use a min-heap to maintain the k largest elements seen so far.
    # The smallest of these k elements will be at the root of the heap.
    min_heap = []
    for num in nums:
        heapq.heappush(min_heap, num)
        # If the heap size exceeds k, remove the smallest element (the root).
        if len(min_heap) > k:
            heapq.heappop(min_heap)
    
    # 3. The root of the heap is the kth largest element.
    return min_heap[0]

def main():
    # Test Case 1
    nums1 = [3, 2, 1, 5, 6, 4]
    k1 = 2
    print(f"Test Case 1: Array = {nums1}, k = {k1} -> Result: {find_kth_largest(nums1, k1)}")

    # Test Case 2
    nums2 = [3, 2, 3, 1, 2, 4, 5, 5, 6]
    k2 = 4
    print(f"Test Case 2: Array = {nums2}, k = {k2} -> Result: {find_kth_largest(nums2, k2)}")

    # Test Case 3
    nums3 = [1]
    k3 = 1
    print(f"Test Case 3: Array = {nums3}, k = {k3} -> Result: {find_kth_largest(nums3, k3)}")

    # Test Case 4
    nums4 = [7, 6, 5, 4, 3, 2, 1]
    k4 = 7
    print(f"Test Case 4: Array = {nums4}, k = {k4} -> Result: {find_kth_largest(nums4, k4)}")

    # Test Case 5
    nums5 = [99, 99]
    k5 = 1
    print(f"Test Case 5: Array = {nums5}, k = {k5} -> Result: {find_kth_largest(nums5, k5)}")

    # Invalid Input Test
    try:
        nums6 = []
        k6 = 1
        find_kth_largest(nums6, k6)
    except ValueError as e:
        print(f"Invalid Input Test: Caught expected exception -> {e}")

if __name__ == "__main__":
    main()

File: dataset/ZeroShot/GEMINI_gemini-2.5-pro/Task147.java
import java.util.PriorityQueue;

public class Task147 {

    /**
     * Finds the kth largest element in an array of integers using a min-heap.
     *
     * @param nums The input array of integers.
     * @param k    The 'k' value, representing the desired largest element rank.
     * @return The kth largest element.
     * @throws IllegalArgumentException if the input is invalid (e.g., null array,
     *                                  empty array, or k is out of bounds).
     */
    public static int findKthLargest(int[] nums, int k) {
        // 1. Input Validation
        if (nums == null || nums.length == 0 || k < 1 || k > nums.length) {
            throw new IllegalArgumentException("Invalid input: array is null/empty or k is out of bounds.");
        }

        // 2. Use a min-heap (PriorityQueue) to maintain the k largest elements seen so far.
        // The smallest of these k elements will be at the root of the heap.
        PriorityQueue<Integer> minHeap = new PriorityQueue<>(k);

        for (int num : nums) {
            minHeap.add(num);
            // If the heap size exceeds k, remove the smallest element (the root).
            if (minHeap.size() > k) {
                minHeap.poll();
            }
        }

        // 3. The root of the heap is the kth largest element.
        return minHeap.peek();
    }

    public static void main(String[] args) {
        // Test Case 1
        int[] nums1 = {3, 2, 1, 5, 6, 4};
        int k1 = 2;
        System.out.println("Test Case 1: Array = [3, 2, 1, 5, 6, 4], k = 2 -> Result: " + findKthLargest(nums1, k1));

        // Test Case 2
        int[] nums2 = {3, 2, 3, 1, 2, 4, 5, 5, 6};
        int k2 = 4;
        System.out.println("Test Case 2: Array = [3, 2, 3, 1, 2, 4, 5, 5, 6], k = 4 -> Result: " + findKthLargest(nums2, k2));

        // Test Case 3
        int[] nums3 = {1};
        int k3 = 1;
        System.out.println("Test Case 3: Array = [1], k = 1 -> Result: " + findKthLargest(nums3, k3));

        // Test Case 4
        int[] nums4 = {7, 6, 5, 4, 3, 2, 1};
        int k4 = 7;
        System.out.println("Test Case 4: Array = [7, 6, 5, 4, 3, 2, 1], k = 7 -> Result: " + findKthLargest(nums4, k4));

        // Test Case 5
        int[] nums5 = {99, 99};
        int k5 = 1;
        System.out.println("Test Case 5: Array = [99, 99], k = 1 -> Result: " + findKthLargest(nums5, k5));

        // Invalid Input Test
        try {
            int[] nums6 = {};
            int k6 = 1;
            findKthLargest(nums6, k6);
        } catch (IllegalArgumentException e) {
            System.out.println("Invalid Input Test: Caught expected exception -> " + e.getMessage());
        }
    }
}